src/rpc/mining.cpp

```diff
@@ -478,9 +478,14 @@ static UniValue getblocktemplate(const JSONRPCRequest& request)
             {
                 if (g_best_block_cv.wait_until(lock, checktxtime) == std::cv_status::timeout)
                 {
-                    // Timeout: Check transactions for update
+                    // Drop the best block mutex while accessing the mempool to avoid deadlock with
+                    // ::mempool.cs.
+                    LEAVE_CRITICAL_SECTION(lock);
+                    // Break if we've seen new transactions after the timeout has elapsed.
                     if (mempool.GetTransactionsUpdated() != nTransactionsUpdatedLastLP)
                         break;
+                    ENTER_CRITICAL_SECTION(lock);
```

Comment:
`ENTER_CRITICAL_SECTION(…);` isn't RAII, right?

Do we need to `LEAVE_CRITICAL_SECTION(…);` when exiting the loop?

Reply:
Nope; the `unlock()`/debug stack pop that `LEAVE_CRITICAL_SECTION(...)` would perform is implicit to the destruction of `lock` when it falls out of scope after the `break`.

Reply:
@jamesob Makes perfect sense: saw the `WAIT_LOCK(g_best_block_mutex, lock);` now. Thanks.